File: cpp/tests/unit/test_bugs_new7.cpp
// Regression tests for BUG-NEW-33 (C++ dead-code removal).
//
// BUG-NEW-33 (C++, §4.4):
//   Dead code was removed from executeInvalidationCommandLocked(): the 9
//   Secure-queue-only commands listed below were unreachable there because
//   processCommand() already intercepts them and writes CERROR_ILL before
//   calling the helper.  The removal changes nothing observable — these tests
//   confirm that CERROR_ILL is STILL raised through processCommand() after
//   the dead code was deleted.
//
//   The 9 commands (all must raise CERROR_ILL on the NS queue):
//     TLBI_EL3_ALL  (§4.1.1 / §4.4.2.9)
//     TLBI_EL3_VA   (§4.1.1 / §4.4.2.9)
//     TLBI_S_EL2_ALL  (§4.4.2.11)
//     TLBI_S_EL2_VA   (§4.4.2.12)
//     TLBI_S_EL2_VAA  (§4.4.2.13)
//     TLBI_S_EL2_ASID (§4.4.2.14)
//     TLBI_S_S12_VMALL (§4.4.3.4)
//     TLBI_S_S2_IPA    (§4.4.3.3)
//     TLBI_SNH_ALL     (§4.4.4.2)
//
// All tests are REGRESSION GUARDS — they should PASS both before and after
// the dead-code removal.
// ─────────────────────────────────────────────────────────────────────────────

#include <gtest/gtest.h>
#include "smmu/smmu.h"
#include "smmu/types.h"
#include "smmu/configuration.h"
#include <cstdint>

using namespace smmu;

// ============================================================================
// Helpers
// ============================================================================

namespace {

static void enableSMMU(SMMU& s) {
    s.setCR0(SMMU::CR0_SMMUEN | SMMU::CR0_EVENTQEN | SMMU::CR0_CMDQEN | SMMU::CR0_PRIQEN);
}

// Build a minimal CommandEntry for the given type.
static CommandEntry makeCmd(CommandType type) {
    CommandEntry cmd;
    cmd.type         = type;
    cmd.cs           = 0u;
    cmd.ssec         = false;
    cmd.streamID     = 0u;
    cmd.pasid        = 0u;
    cmd.asid         = 0u;
    cmd.vmid         = 0u;
    cmd.startAddress = 0u;
    return cmd;
}

// BUG-NEW-40 fix: getGerror() already returns gerrorStatus XOR gerrorNStatus
// (the active error bits).  The former double-XOR recovered the raw gerrorStatus
// instead of the active-error mask.  One mask is sufficient.
static bool isGerrorCmdqErrActive(const SMMU& s) {
    return (s.getGerror() & GERROR_CMDQ_ERR) != 0u;
}

} // anonymous namespace

// ============================================================================
// BUG-NEW-33: dead-code removal regression guards
// ============================================================================
// All nine tests follow the same structure:
//   1. Enable SMMU.
//   2. Submit the Secure-queue-only command to the NS queue.
//   3. Process the command queue.
//   4. Verify that CMDQ_CONS.ERR == CERROR_ILL and GERROR.CMDQ_ERR is active.
//
// These are REGRESSION GUARDS — they must PASS both before and after the
// dead-code removal.

// Test 1: TLBI_EL3_ALL on NS queue must still raise CERROR_ILL (§4.4.2.9).
TEST(BugNew33DeadCodeRegression, El3All_Still_Raises_CERROR_ILL) {
    SMMU s;
    enableSMMU(s);

    CommandEntry cmd = makeCmd(CommandType::TLBI_EL3_ALL);
    ASSERT_TRUE(s.submitCommand(cmd).isOk());
    s.processCommandQueue();

    EXPECT_EQ(s.getCmdqConsErr(), CERROR_ILL)
        << "TLBI_EL3_ALL must still write CERROR_ILL after dead-code removal (BUG-NEW-33)";
    EXPECT_TRUE(isGerrorCmdqErrActive(s))
        << "TLBI_EL3_ALL must still raise GERROR.CMDQ_ERR after dead-code removal (BUG-NEW-33)";
}

// Test 2: TLBI_EL3_VA on NS queue must still raise CERROR_ILL (§4.4.2.9).
TEST(BugNew33DeadCodeRegression, El3Va_Still_Raises_CERROR_ILL) {
    SMMU s;
    enableSMMU(s);

    CommandEntry cmd = makeCmd(CommandType::TLBI_EL3_VA);
    cmd.startAddress = 0x0000'1000u;
    ASSERT_TRUE(s.submitCommand(cmd).isOk());
    s.processCommandQueue();

    EXPECT_EQ(s.getCmdqConsErr(), CERROR_ILL)
        << "TLBI_EL3_VA must still write CERROR_ILL after dead-code removal (BUG-NEW-33)";
    EXPECT_TRUE(isGerrorCmdqErrActive(s))
        << "TLBI_EL3_VA must still raise GERROR.CMDQ_ERR after dead-code removal (BUG-NEW-33)";
}

// Test 3: TLBI_S_EL2_ALL on NS queue must still raise CERROR_ILL (§4.4.2.11).
TEST(BugNew33DeadCodeRegression, SEL2All_Still_Raises_CERROR_ILL) {
    SMMU s;
    enableSMMU(s);

    CommandEntry cmd = makeCmd(CommandType::TLBI_S_EL2_ALL);
    ASSERT_TRUE(s.submitCommand(cmd).isOk());
    s.processCommandQueue();

    EXPECT_EQ(s.getCmdqConsErr(), CERROR_ILL)
        << "TLBI_S_EL2_ALL must still write CERROR_ILL after dead-code removal (BUG-NEW-33)";
    EXPECT_TRUE(isGerrorCmdqErrActive(s))
        << "TLBI_S_EL2_ALL must still raise GERROR.CMDQ_ERR after dead-code removal (BUG-NEW-33)";
}

// Test 4: TLBI_S_EL2_VA on NS queue must still raise CERROR_ILL (§4.4.2.12).
TEST(BugNew33DeadCodeRegression, SEL2Va_Still_Raises_CERROR_ILL) {
    SMMU s;
    enableSMMU(s);

    CommandEntry cmd = makeCmd(CommandType::TLBI_S_EL2_VA);
    cmd.startAddress = 0x0000'2000u;
    ASSERT_TRUE(s.submitCommand(cmd).isOk());
    s.processCommandQueue();

    EXPECT_EQ(s.getCmdqConsErr(), CERROR_ILL)
        << "TLBI_S_EL2_VA must still write CERROR_ILL after dead-code removal (BUG-NEW-33)";
    EXPECT_TRUE(isGerrorCmdqErrActive(s))
        << "TLBI_S_EL2_VA must still raise GERROR.CMDQ_ERR after dead-code removal (BUG-NEW-33)";
}

// Test 5: TLBI_S_EL2_VAA on NS queue must still raise CERROR_ILL (§4.4.2.13).
TEST(BugNew33DeadCodeRegression, SEL2Vaa_Still_Raises_CERROR_ILL) {
    SMMU s;
    enableSMMU(s);

    CommandEntry cmd = makeCmd(CommandType::TLBI_S_EL2_VAA);
    cmd.startAddress = 0x0000'3000u;
    ASSERT_TRUE(s.submitCommand(cmd).isOk());
    s.processCommandQueue();

    EXPECT_EQ(s.getCmdqConsErr(), CERROR_ILL)
        << "TLBI_S_EL2_VAA must still write CERROR_ILL after dead-code removal (BUG-NEW-33)";
    EXPECT_TRUE(isGerrorCmdqErrActive(s))
        << "TLBI_S_EL2_VAA must still raise GERROR.CMDQ_ERR after dead-code removal (BUG-NEW-33)";
}

// Test 6: TLBI_S_EL2_ASID on NS queue must still raise CERROR_ILL (§4.4.2.14).
TEST(BugNew33DeadCodeRegression, SEL2Asid_Still_Raises_CERROR_ILL) {
    SMMU s;
    enableSMMU(s);

    CommandEntry cmd = makeCmd(CommandType::TLBI_S_EL2_ASID);
    cmd.asid = 0u;
    ASSERT_TRUE(s.submitCommand(cmd).isOk());
    s.processCommandQueue();

    EXPECT_EQ(s.getCmdqConsErr(), CERROR_ILL)
        << "TLBI_S_EL2_ASID must still write CERROR_ILL after dead-code removal (BUG-NEW-33)";
    EXPECT_TRUE(isGerrorCmdqErrActive(s))
        << "TLBI_S_EL2_ASID must still raise GERROR.CMDQ_ERR after dead-code removal (BUG-NEW-33)";
}

// Test 7: TLBI_S_S12_VMALL on NS queue must still raise CERROR_ILL (§4.4.3.4).
TEST(BugNew33DeadCodeRegression, SS12Vmall_Still_Raises_CERROR_ILL) {
    SMMU s;
    enableSMMU(s);

    CommandEntry cmd = makeCmd(CommandType::TLBI_S_S12_VMALL);
    ASSERT_TRUE(s.submitCommand(cmd).isOk());
    s.processCommandQueue();

    EXPECT_EQ(s.getCmdqConsErr(), CERROR_ILL)
        << "TLBI_S_S12_VMALL must still write CERROR_ILL after dead-code removal (BUG-NEW-33)";
    EXPECT_TRUE(isGerrorCmdqErrActive(s))
        << "TLBI_S_S12_VMALL must still raise GERROR.CMDQ_ERR after dead-code removal (BUG-NEW-33)";
}

// Test 8: TLBI_S_S2_IPA on NS queue must still raise CERROR_ILL (§4.4.3.3).
TEST(BugNew33DeadCodeRegression, SS2Ipa_Still_Raises_CERROR_ILL) {
    SMMU s;
    enableSMMU(s);

    CommandEntry cmd = makeCmd(CommandType::TLBI_S_S2_IPA);
    cmd.startAddress = 0x0000'4000u;
    ASSERT_TRUE(s.submitCommand(cmd).isOk());
    s.processCommandQueue();

    EXPECT_EQ(s.getCmdqConsErr(), CERROR_ILL)
        << "TLBI_S_S2_IPA must still write CERROR_ILL after dead-code removal (BUG-NEW-33)";
    EXPECT_TRUE(isGerrorCmdqErrActive(s))
        << "TLBI_S_S2_IPA must still raise GERROR.CMDQ_ERR after dead-code removal (BUG-NEW-33)";
}

// Test 9: TLBI_SNH_ALL on NS queue must still raise CERROR_ILL (§4.4.4.2).
TEST(BugNew33DeadCodeRegression, SnhAll_Still_Raises_CERROR_ILL) {
    SMMU s;
    enableSMMU(s);

    CommandEntry cmd = makeCmd(CommandType::TLBI_SNH_ALL);
    ASSERT_TRUE(s.submitCommand(cmd).isOk());
    s.processCommandQueue();

    EXPECT_EQ(s.getCmdqConsErr(), CERROR_ILL)
        << "TLBI_SNH_ALL must still write CERROR_ILL after dead-code removal (BUG-NEW-33)";
    EXPECT_TRUE(isGerrorCmdqErrActive(s))
        << "TLBI_SNH_ALL must still raise GERROR.CMDQ_ERR after dead-code removal (BUG-NEW-33)";
}

File: rust/smmu/tests/test_bugs_new7.rs
//! Regression tests for BUG-NEW-34 and BUG-NEW-35 (Rust).
//!
//! # BUG-NEW-34 (Rust, §4.4.2.9)
//!
//! In `process_single_command()`, the `TlbiEl3All` and `TlbiEl3Va` match arms
//! omit the inline `signal_gerror(GERROR_CMDQ_ERR)` call that every other
//! CERROR_ILL path includes.  `process_command_queue()` compensates by calling
//! `signal_gerror` in the caller after detecting a CERROR.  The observable
//! behaviour is identical before and after any refactor that moves the call
//! into the arm.  These tests confirm that `GERROR.CMDQ_ERR` is set after
//! submitting and processing each command.
//!
//! # BUG-NEW-35 (Rust, §4.1.1)
//!
//! Same pattern as BUG-NEW-34 for `DptiAll` and `DptiPa`.  Both commands
//! are Secure-only / unimplemented in this model and must raise CERROR_ILL.
//! The inline `signal_gerror` call is missing from their match arms;
//! `process_command_queue()` compensates.  Again, observable behaviour is
//! unchanged.
//!
//! # BUG-NEW-36 (Both)
//!
//! Comment-only fix — no test needed.
//!
//! # Test classification
//!
//! All tests are REGRESSION GUARDS.  They should PASS both before and after
//! any refactor that moves the `signal_gerror` call into the individual arms.

#![allow(missing_docs)]
#![allow(clippy::unwrap_used)]
#![allow(clippy::items_after_statements)]
#![allow(clippy::doc_markdown)]
#![allow(clippy::uninlined_format_args)]

use smmu::types::{CommandEntry, CommandType};
use smmu::SMMU;

// ============================================================================
// Helpers
// ============================================================================

/// Build an SMMU with SMMUEN + CMDQEN + EVENTQEN + PRIQEN enabled.
fn make_smmu() -> SMMU {
    let smmu = SMMU::new();
    smmu.set_cr0(SMMU::CR0_SMMUEN | SMMU::CR0_CMDQEN | SMMU::CR0_EVENTQEN | SMMU::CR0_PRIQEN);
    smmu
}

/// Returns true when GERROR.CMDQ_ERR is active (GERROR XOR GERRORN has bit 0 set).
fn is_gerror_cmdq_err_active(smmu: &SMMU) -> bool {
    (smmu.get_gerror() ^ smmu.get_gerrorn()) & SMMU::GERROR_CMDQ_ERR != 0
}

// ============================================================================
// BUG-NEW-34: TlbiEl3All and TlbiEl3Va — confirm GERROR.CMDQ_ERR is set
// ============================================================================
// ARM §4.4.2.9: CMD_TLBI_EL3_ALL and CMD_TLBI_EL3_VA are valid only on the
// Secure Command queue; on the Non-secure Command queue they cause CERROR_ILL.
//
// These tests are REGRESSION GUARDS — they must PASS both before and after
// any refactor that moves the signal_gerror call into the match arm.

/// BUG-NEW-34 (regression guard): `TlbiEl3All` on NS queue must set
/// GERROR.CMDQ_ERR after processing.
///
/// ARM §4.4.2.9: valid only on Secure queue; NS queue → CERROR_ILL.
/// The inline `signal_gerror` call is missing from this arm; the caller
/// compensates.  Observable behaviour is unchanged — this guard confirms it.
#[test]
fn bug_new34_tlbi_el3_all_gerror_cmdq_err_active() {
    let smmu = make_smmu();

    let cmd = CommandEntry::new(CommandType::TlbiEl3All, 0, 0);
    smmu.submit_command(cmd).unwrap();
    let _ = smmu.process_command_queue();

    assert_eq!(
        smmu.get_cmdq_cons_err(),
        SMMU::CERROR_ILL,
        "BUG-NEW-34: TlbiEl3All on NS queue must write CERROR_ILL \
         (ARM §4.4.2.9). Got CERROR={}",
        smmu.get_cmdq_cons_err()
    );
    assert!(
        is_gerror_cmdq_err_active(&smmu),
        "BUG-NEW-34: TlbiEl3All on NS queue must assert GERROR.CMDQ_ERR \
         (ARM §4.4.2.9)"
    );
}

/// BUG-NEW-34 (regression guard): `TlbiEl3Va` on NS queue must set
/// GERROR.CMDQ_ERR after processing.
///
/// ARM §4.4.2.9: valid only on Secure queue; NS queue → CERROR_ILL.
#[test]
fn bug_new34_tlbi_el3_va_gerror_cmdq_err_active() {
    let smmu = make_smmu();

    let mut cmd = CommandEntry::new(CommandType::TlbiEl3Va, 0, 0);
    cmd.start_address = 0x0000_1000u64;
    smmu.submit_command(cmd).unwrap();
    let _ = smmu.process_command_queue();

    assert_eq!(
        smmu.get_cmdq_cons_err(),
        SMMU::CERROR_ILL,
        "BUG-NEW-34: TlbiEl3Va on NS queue must write CERROR_ILL \
         (ARM §4.4.2.9). Got CERROR={}",
        smmu.get_cmdq_cons_err()
    );
    assert!(
        is_gerror_cmdq_err_active(&smmu),
        "BUG-NEW-34: TlbiEl3Va on NS queue must assert GERROR.CMDQ_ERR \
         (ARM §4.4.2.9)"
    );
}

// ============================================================================
// BUG-NEW-35: DptiAll and DptiPa — confirm GERROR.CMDQ_ERR is set
// ============================================================================
// ARM §4.1.1: CMD_DPTI_ALL (opcode 0x70) and CMD_DPTI_PA (opcode 0x73) are
// dirty-page tracking invalidation commands.  This model does not implement
// dirty-page tracking, so both must raise CERROR_ILL when submitted.
//
// Same pattern as BUG-NEW-34: the inline `signal_gerror` call is missing from
// the match arms; the caller compensates.  Observable behaviour is unchanged.
//
// These tests are REGRESSION GUARDS — they must PASS both before and after
// any refactor that moves the signal_gerror call into the match arm.

/// BUG-NEW-35 (regression guard): `DptiAll` must set GERROR.CMDQ_ERR after
/// processing.
///
/// This model does not implement dirty-page tracking invalidation; the command
/// must raise CERROR_ILL on any queue submission.
#[test]
fn bug_new35_dpti_all_gerror_cmdq_err_active() {
    let smmu = make_smmu();

    let cmd = CommandEntry::new(CommandType::DptiAll, 0, 0);
    smmu.submit_command(cmd).unwrap();
    let _ = smmu.process_command_queue();

    assert_eq!(
        smmu.get_cmdq_cons_err(),
        SMMU::CERROR_ILL,
        "BUG-NEW-35: DptiAll must write CERROR_ILL \
         (dirty-page tracking not implemented). Got CERROR={}",
        smmu.get_cmdq_cons_err()
    );
    assert!(
        is_gerror_cmdq_err_active(&smmu),
        "BUG-NEW-35: DptiAll must assert GERROR.CMDQ_ERR \
         (dirty-page tracking not implemented)"
    );
}

/// BUG-NEW-35 (regression guard): `DptiPa` must set GERROR.CMDQ_ERR after
/// processing.
///
/// This model does not implement dirty-page tracking invalidation; the command
/// must raise CERROR_ILL on any queue submission.
#[test]
fn bug_new35_dpti_pa_gerror_cmdq_err_active() {
    let smmu = make_smmu();

    let mut cmd = CommandEntry::new(CommandType::DptiPa, 0, 0);
    cmd.start_address = 0u64;
    smmu.submit_command(cmd).unwrap();
    let _ = smmu.process_command_queue();

    assert_eq!(
        smmu.get_cmdq_cons_err(),
        SMMU::CERROR_ILL,
        "BUG-NEW-35: DptiPa must write CERROR_ILL \
         (dirty-page tracking not implemented). Got CERROR={}",
        smmu.get_cmdq_cons_err()
    );
    assert!(
        is_gerror_cmdq_err_active(&smmu),
        "BUG-NEW-35: DptiPa must assert GERROR.CMDQ_ERR \
         (dirty-page tracking not implemented)"
    );
}
